box/src/main/scala/blobstore/box/BoxStore.scala

```diff
@@ -0,0 +1,238 @@
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+package blobstore
+package box
+
+
+import java.io.{InputStream, OutputStream, PipedInputStream, PipedOutputStream}
+
+import cats.effect.Effect
+import com.box.sdk.{BoxAPIConnection, BoxFile, BoxFolder, BoxItem}
+import fs2.{Sink, Stream}
+
+import scala.collection.JavaConverters._
+import scala.concurrent.ExecutionContext
+
+case class BoxStore[F[_]](api: BoxAPIConnection, rootFolderId: String)(implicit F: Effect[F], ec: ExecutionContext) extends Store[F] {
+
+  val rootFolder = new BoxFolder(api, rootFolderId)
+
+  def boxItemAtPath(path: Path): Option[BoxItem] = {
+    boxItemAtPath(rootFolder, path.root :: path.key.split("/").toList)
+  }
+
+  def boxItemAtPath(parentFolder: BoxFolder, pathParts: List[String]): Option[BoxItem] = {
+    pathParts match {
+      case Nil => None
+      case head::Nil =>
+        parentFolder.getChildren.asScala.find(info => info.getName.equalsIgnoreCase(head))
+          .map(_.getResource.asInstanceOf[BoxItem])
+      case head::tail =>
+        parentFolder.getChildren.asScala.find(info => info.getName.equalsIgnoreCase(head))
+          .flatMap(info => boxItemAtPath(new BoxFolder(api, info.getID), tail))
+    }
+  }
+
+  def boxFileAtPath(path: Path): Option[BoxFile] = {
+    boxItemAtPath(path).flatMap {
+      case file: BoxFile => Some(file)
+      case _ => None
+    }
+  }
+
+  def boxFolderAtPath(path: Path): Option[BoxFolder] = {
+    boxItemAtPath(path).flatMap {
+      case folder: BoxFolder => Some(folder)
+      case _ => None
+    }
+  }
+
+
+  /**
+    * List paths. See [[StoreOps.ListOps]] for convenient listAll method.
+    *
+    * @param path to list
+    * @return stream of Paths. Implementing stores must guarantee that returned Paths
+    *         have correct values for size, isDir and lastModified.
+    */
+  override def list(path: Path): fs2.Stream[F, Path] = {
+    for {
+      itemOpt <- Stream.eval(F.delay(boxItemAtPath(path)))
+      item <- if (itemOpt.isDefined) Stream.emit(itemOpt.get) else Stream.empty
+
+      listFiles = Stream.eval(F.delay(item.asInstanceOf[BoxFolder]))
+          .flatMap(folder => Stream.fromIterator(folder.getChildren.iterator.asScala))
+          .map(itemInfo => {
+            val isDir = itemInfo.isInstanceOf[BoxFolder#Info]
+            val pathKey = if (path.key.takeRight(1).equals("/")) path.key.dropRight(1) else path.key
+            val childKey = s"$pathKey/${itemInfo.getName}"
+            Path(path.root, childKey, Some(itemInfo.getSize), isDir, Some(itemInfo.getModifiedAt))
+          })
+
+      listOneFile = Stream.eval(F.delay(
+        Path(path.root, path.key, Some(item.getInfo.getSize), isDir = false, Some(item.getInfo.getModifiedAt))
+      ))
+
+      s <- if (item.isInstanceOf[BoxFolder]) listFiles else listOneFile
```

Comment:
It seems like we could check this condition without having to construct the listOneFile stream if it's met
cc: @rolandomanrique 

But maybe something like

``
item.isInstanceOf[BoxFolder].ifM(listFiles, Stream.eval(F.delay { ... }))
``

Reply:
I can't imagine it takes a significant amount of time to construct that stream. Btw we are doing the same thing in `FileStore.list`

Reply:
You would need a Stream to do `ifM` any ways.